### foot specific ###
############ Differential Methylation Analysis ############

```{r setup, include=FALSE}
### Load libraries ###
library(vegan)
library(edgeR)
library(tidyverse)
library(SYNCSA)
library(mice)
library(ape)
library(rtracklayer)
library(genomation)
library(plyranges)
library(GenomicRanges)
library(lme4)
library(emmeans)
library(ggthemes)

```

```{r}
### Set working directory, include all (top5 for each treatments) foot coverage files ###
setwd("/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/cg_coverage_files/tissue_specific/foot_final")

### Load metadata file that include sample names, load all coverage files into R ###
meta_data_foot<-read.delim("foot_final_metadata.txt", row.names = "sample", stringsAsFactors = FALSE)
Sample_foot <- row.names(meta_data_foot)
files_foot <- paste0(Sample_foot,".CpG_report.merged_CpG_evidence.cov.CpG_report.merged_CpG_evidence.cov")

### EdgeR function readBismark2DGE reads all the files and collates the counts for all the sample into one data object ###
yall <- readBismark2DGE(files_foot, sample.names=Sample_foot)

### Check dimension and the count matrix ###
dim(yall)
head(yall$counts)

### Save a data frame for downstream analyses ###
yall_df<-as.data.frame(yall)

### Sum up the counts of methylated and unmethylated reads to get the total read coverage at each CpG site for each sample ###
Methylation <- gl(2, 1, ncol(yall), labels = c("Me", "Un"))
Me <- yall$counts[ , Methylation == "Me" ]
Un <- yall$counts[ , Methylation == "Un" ]
Coverage <- Me + Un
# Prefiltered total # of CpGs sequenced: colSums(Coverage > 0, na.rm = TRUE) 
head(Coverage)

### Calculating prefiltered genome-wide coverage - Replace zeros with NA so they're excluded in mean
# Coverage[Coverage == 0] <- NA
# # Compute mean per column, excluding NA (formerly zero)
# mean_coverage <- colMeans(Coverage, na.rm = TRUE)
# mean_coverage

### Calculating average genome wide methylation percentage prefiltering
# prop_meth_matrix_foot_prefilter <- Me /(Un+Me)
# colMeans(prop_meth_matrix_foot_prefilter,na.rm=TRUE)*100

## Calculating # of unique CpGs 
# covered_per_cpg <- rowSums(Coverage > 0, na.rm = TRUE)
# sum(covered_per_cpg > 0) # 1946085

### Filtering to only include samples with a coverage of at least 3, for 66% of samples, total of 14 samples ###
n=3
keep_foot <- rowSums(Coverage >= n) >= 14
table(keep_foot)

### DGEList object is subsetted to retain only the filtered loci ###
y_foot <- yall[keep_foot,, keep.lib.sizes=FALSE]

### Normalization - set the library sizes for each sample to be the average of the total read counts for the methylated and unmethylated libraries ###
TotalLibSize <- y_foot$samples$lib.size[Methylation=="Me"] +
  +                 y_foot$samples$lib.size[Methylation=="Un"]
y_foot$samples$lib.size <- rep(TotalLibSize, each=2)
y_foot$samples

### Compute the corresponding methylation summary from the methylated and unmethylated counts ###
Me_foot <- y_foot$counts[, Methylation=="Me"]
Un_foot <- y_foot$counts[, Methylation=="Un"]

### Calculating prefiltered genome-wide coverage 
# Coverage_postfilter<-Un_foot + Me_foot
# Coverage_postfilter[Coverage_postfilter == 0] <- NA
# mean_coverage_postfilter <- colMeans(Coverage_postfilter, na.rm = TRUE)
# mean_coverage_postfilter 

### Calculating a methylation proportion matrix
prop_meth_matrix_foot <- Me_foot/(Me_foot+Un_foot)
#Postfiltered total # of CpGs: colSums(!is.na(prop_meth_matrix_foot))

# ### Calculating average genome wide methylation percentage postfiltering
# colMeans(prop_meth_matrix_foot,na.rm=TRUE)*100

```


```{r}

### We want to use the exposed site as reference level for both transplant and origin site effect analyses, need to relevel for transplant site effect ###
meta_data_foot$transplant_site <- relevel(factor(meta_data_foot$transplant_site), ref = "exposed")
meta_data_foot$origin_site <- relevel(factor(meta_data_foot$origin_site), ref = "exposed")

### Create a design matrix, using origin site, transplant site, and final shell length as fixed effects ###
designSL_foot <- model.matrix(~0 + origin_site + transplant_site +
                           LENGTH_FINAL..mm., 
                           data=meta_data_foot)

### Expand to the full design matrix modeling the sample and methylation effects ###
design_foot <- modelMatrixMeth(designSL_foot)

### Dispersion estimation ###
y_foot <- estimateDisp(y_foot, design = design_foot, robust = TRUE)

### Create the BCV plot for visualization ### 
plotBCV(y_foot)

### Testing for differentially methylated CpG loci ###

### fit NB GLMs for all the CpG loci using the glmFit function in edgeR ###
fit_foot <- glmFit(y_foot, design_foot)

### Testing for differentially methylated CpG sites between different treatment groups using likelihood ratio tests ###

### Origin site effects ###
contr_origin_foot <- makeContrasts(Origin = origin_siteprotected-origin_siteexposed, levels = design_foot)
lrt_origin_foot <- glmLRT(fit_foot, contrast=contr_origin_foot)
summary( decideTests(lrt_origin_foot) )

### Transplant site effects (exposed is the reference level) ###
contr_trans_foot <- makeContrasts(Transplant = transplant_siteprotected, levels = design_foot)
lrt_trans_foot <- glmLRT(fit_foot, contrast=contr_trans_foot)
summary( decideTests(lrt_trans_foot) )

### Length effect ###
contr_length_foot <- makeContrasts(Origin = LENGTH_FINAL..mm., levels = design_foot)
lrt_length_foot <- glmLRT(fit_foot, contrast=contr_length_foot)
summary( decideTests(lrt_length_foot) )

### Wrangle data for generating volano plot of CpG DM ###
### Correct p-values using BH method ###
lrt_origin_foot$table$FDR <- p.adjust( lrt_origin_foot$table$PValue, method = "BH" )
lrt_trans_foot$table$FDR <- p.adjust( lrt_trans_foot$table$PValue, method = "BH" )
lrt_length_foot$table$FDR <- p.adjust( lrt_length_foot$table$PValue, method = "BH" )

### Apply logical variable for significant DM ###
lrt_origin_foot$table$Sig <- ifelse(lrt_origin_foot$table$FDR< 0.05, TRUE, FALSE )
lrt_trans_foot$table$Sig <- ifelse(lrt_trans_foot$table$FDR< 0.05, TRUE, FALSE )
lrt_length_foot$table$Sig <- ifelse(lrt_length_foot$table$FDR < 0.05, TRUE, FALSE )

### Apply binary variable for hyper/hypomethylation or "Up" vs "Down" ###
lrt_origin_foot$table$Dir <- ifelse(lrt_origin_foot$table$logFC > 0, "Up", "Down" )
lrt_trans_foot$table$Dir <- ifelse( lrt_trans_foot$table$logFC > 0, "Up", "Down" )
lrt_length_foot$table$Dir <- ifelse(lrt_length_foot$table$logFC > 0, "Up", "Down")

### Create combined term for significance and fold-change direction of diff meth ###
lrt_origin_foot$table$Sig_Dir <- paste( lrt_origin_foot$table$Sig,
                             lrt_origin_foot$table$Dir,
                             sep = "_" )
lrt_trans_foot$table$Sig_Dir <- paste( lrt_trans_foot$table$Sig,
                             lrt_trans_foot$table$Dir,
                             sep = "_" )
lrt_length_foot$table$Sig_Dir <- paste( lrt_length_foot$table$Sig,
                             lrt_length_foot$table$Dir,
                             sep = "_" )

### Create CpG ID and treatment variables ###
lrt_origin_foot$table$Treat <- "Origin"
lrt_trans_foot$table$Treat <- "Transplant"
lrt_length_foot$table$Treat <- "Length"

lrt_origin_foot$table$sample <- rownames(lrt_origin_foot$table)
lrt_trans_foot$table$sample <- rownames(lrt_trans_foot$table)
lrt_length_foot$table$sample <- rownames(lrt_length_foot$table)


### Run Wilcoxon test to see if there's any skewness of DM 
wilcox.test(lrt_trans_foot$table$logFC, mu = 0, alternative = "less")  #p-value < 2.2e-16
wilcox.test(lrt_trans_foot$table$logFC, mu = 0, alternative = "greater")   

wilcox.test(lrt_origin_foot$table$logFC, mu = 0, alternative = "less") 
wilcox.test(lrt_origin_foot$table$logFC, mu = 0, alternative = "greater")  

### Merge origin site and transplant site coefficients ###
all_CpG_dm_foot <- rbind( lrt_origin_foot$table,
                     lrt_trans_foot$table )

## with length effect
all_CpG_dm_length <- rbind( lrt_origin_foot$table,
                     lrt_trans_foot$table,
                     lrt_length_foot$table)

## Getting a full table of unique CpGs for plotting the relationship of all log2FC between gill and foot
# 
# transplant_cpg_foot<- all_CpG_dm_foot  %>%
#   filter(Treat=="Transplant")

# write_csv(all_CpG_dm_foot ,"/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/manuscript/foot_unique_all_CpG.csv")


### Saving DM sites for transplant site effect and origin site effect as new objects ###
DM_trans_foot<- lrt_trans_foot$table
DM_origin_foot<- lrt_origin_foot$table
DM_length_foot<- lrt_length_foot$table

DM_Trans_foot<-DM_trans_foot%>% filter(Sig =="TRUE",Treat =="Transplant")
DM_Origin_foot<-DM_origin_foot %>% filter(Sig =="TRUE", Treat =="Origin")
DM_length_foot<-DM_length_foot%>% filter(Sig =="TRUE", Treat =="Length")

####### For plotting ########

library(patchwork)
# Filter datasets
plot_origin <- all_CpG_dm_foot %>% filter(Treat == "Origin")
plot_trans <- all_CpG_dm_foot %>% filter(Treat == "Transplant")
  
# Individual plots
p1_foot <- ggplot(plot_trans, aes(x = logFC, y = -log10(PValue), color = Sig_Dir)) +
  geom_point(size = 8, alpha =0.4, stroke=0) +
  scale_color_manual(values = c("#2a2e2e", "#2a2e2e", "blue", "red")) +
  labs(x = "log2FC", y = "-log10 p-value", title = "Transplant",color="black") +
  theme_few(base_size = 28) +
  theme(legend.position = "none",
        plot.title = element_blank(),
   axis.text = element_text(color = "black"),
    axis.ticks = element_line(color = "black"),
    axis.title = element_text(color = "black")
   )

p1_foot 

p2_foot <- ggplot(plot_origin, aes(x = logFC, y = -log10(PValue), color = Sig_Dir))+
  geom_point(size = 8, alpha =0.4,stroke=0) +
  scale_color_manual(values = c("#2a2e2e", "#2a2e2e", "blue", "red")) +
  labs(x = "", y = "-log10 p-value", title = "Origin", color="black") +
  theme_few(base_size = 28) +
  theme(legend.position = "none",
        plot.title = element_blank(),
   axis.text = element_text(color = "black"),
    axis.ticks = element_line(color = "black"),
    axis.title = element_text(color = "black"))
p2_foot 

common_xlim <- c(-12, 12)
common_ylim <- c(0, 10)
# aspect_ratio_value <- 0.6  # Adjust this value as needed (1 means square)

p1_foot <- p1_foot + coord_cartesian(xlim = common_xlim, ylim = common_ylim)
p2_foot <- p2_foot + coord_cartesian(xlim = common_xlim, ylim = common_ylim)

# Stack vertically with shared layout
foot_volcano<-p2_foot / p1_foot  # This uses patchwork
foot_volcano

#ggsave("foot_volcano.png", foot_volcano, width = 10, height = 12, dpi = 300)

```



```{r}
### Find genomics regions that each (DM) CpG falls into ###
### Read in intron annotated gff, see Preprocess_00/add_intron.sh for code for intron annotation ###
gff_intron<-read.gff("/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/cg_coverage_files/TOP_5/new_genomic_intron.gff")

### Convert the gff object to a GRanges object using the gffToGRanges() function of the R package genomation ###
all_GRange<-gffToGRanges("/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/cg_coverage_files/TOP_5/new_genomic_intron.gff", filter = NULL, zero.based = FALSE, ensembl = FALSE) 

### Divide the annotation GRanges object into different objects filtered for exons and introns ###
exon_GRange<-all_GRange %>% 
  filter(type == c
         ("exon"))

intron_GRange<-all_GRange %>% 
  filter(type == c
         ("intron")) 
         
### Code to find PROMOTER REGIONS: defined as 1kb downstream or upstream of first exons of all genes ###

### Separate positive and negative strand of all exons ###
exon_pos_strand <- exon_GRange %>% filter(strand == "+")
exon_neg_strand <- exon_GRange %>% filter(strand == "-")

### Extracting all first exons of each gene ###
### For positive strands, get the first exons by extracting the smallest start position with slice_min ###
first_exon_pos <- as.data.frame(exon_pos_strand) %>% 
  group_by(gene) %>% 
  slice_min(order_by = start, n = 1, with_ties =FALSE) # Only return one when there are overlapping first exons to avoid duplicates

### For negative strands, extract the first exons using the largest start position with slice_max, (going from right to left) ###
first_exon_neg <- as.data.frame(exon_neg_strand) %>% 
  group_by(gene) %>% 
  slice_max(order_by = start, n = 1, with_ties =FALSE)

### Combine both positive and negative strand first exons ###
first_exon_GRange <- bind_rows(first_exon_pos, first_exon_neg)

### Remove potential duplicated lines ###
first_exon_GRange <- first_exon_GRange %>% distinct()

### Create a dataframe ###
first_exon_GRange_df<-as.data.frame(first_exon_GRange)

### Create new GRange object with only first exons ###
first_exon_GRange <- GRanges(
  seqnames = Rle(first_exon_GRange$seqnames),
  ranges = IRanges(start = first_exon_GRange$start, end = first_exon_GRange$end),
  strand = first_exon_GRange$strand,
  gene_id = first_exon_GRange$gene # genes assigned to promoters are those that are associated with the first exons
)

### For some first exons start site <1000, 1kb upstream returns negative values, adjust length that the exons with start position less than 1000 has a start site of 1  

### Adjust start positions for promoter regions based on strand ###
adjusted_start <- ifelse(strand(first_exon_GRange) == "+",
                         start(first_exon_GRange) - 1001,  # Positive strand: promoter starts from upstream 1kb, not counting the first position of exon
                         end(first_exon_GRange)+1)    # Negative strand: end of first exon (from left to right), add 1 to avoid overlap with exon

### Ensure the adjusted start positions are not less than 1 ###
adjusted_start <- pmax(adjusted_start, 1)

# Adjust end positions for promoters based on strand and start conditions
adjusted_end <- ifelse(
  strand(first_exon_GRange) == "+" & start(first_exon_GRange) == 1,
  1000,  # If positive strand and start is 1, end at 1000
  ifelse(
    strand(first_exon_GRange) == "+",
    start(first_exon_GRange) - 1,  # For positive strand, end 1 bp before the first exon to avoid overlap
    end(first_exon_GRange)+1001  # For negative strand, end 1000 downstream of first exon
  )
)

### Create the GRanges object for promoter regions ###
promoters_Grange<- GRanges(
  seqnames = seqnames(first_exon_GRange),
  ranges = IRanges(
    start = adjusted_start,
    end = adjusted_end
  ),
  strand = strand(first_exon_GRange),
  gene = mcols(first_exon_GRange)$gene_id 
)

### Save promoter GRange as new dataframe ###
promoters_Grange_df<-as.data.frame(promoters_Grange)


### Convert a data frame of coordinates of CpGs to a GRange object using the function makeGRangesFromDataFrame() of the R package GenomicRanges ###

#### Use prop_meth_matrix from above, extract first rows as all the CpG positions across samples ###
prop_meth_matrix_foot <- as.data.frame(prop_meth_matrix_foot)

### Reformating ###
prop_meth_matrix_foot_new <- prop_meth_matrix_foot %>%
  rownames_to_column(var = "name") #change row names to column called name

CpGs_foot <- prop_meth_matrix_foot_new  %>%
  separate(col = 1, into = c("seqid", "start"), sep = "-") #separating chromosome, start site information

CpGs_foot$start<-as.numeric(CpGs_foot$start)

CpGs_foot<-CpGs_foot %>% 
  mutate(start= start+1,
         end =start+1)  #changing from 0 to 1 base format because the annotation file is 1 base, but the coverage files are 0 base format.

### Reorder columns to place 'end' in the third column for easier viewing ###
CpGs_foot <- CpGs_foot[, c(1, 2, ncol(CpGs_foot), 3:(ncol(CpGs_foot)-1))]

### Make cpg a GRange object for all CpG sites ###
CpGs_foot<-makeGRangesFromDataFrame(CpGs_foot)
df_cpg_foot<-as.data.frame(CpGs_foot) ### save a new dataframe for all CpGs ###

### Subsetting CpG GRange objects according to genic features ###
### Exon subset ###
exon_subset_foot<-subsetByOverlaps(CpGs_foot, exon_GRange)
#exon_subset<-as.data.frame(exon_subset) 

### Adding gene id to each entry ###
exon_gff <- subset(all_GRange, type == "exon") #need to be GRange object to subset
exon_id_overlaps_foot <- findOverlaps(exon_subset_foot, exon_gff)

### Extract gene_id for matching exons from the GFF file ###
gene_ids_foot <- mcols(exon_gff)$gene[subjectHits(exon_id_overlaps_foot)]
query_hits_foot <- queryHits(exon_id_overlaps_foot)
# Add gene_id only to those rows in exon_subset that have corresponding gene overlaps

### Initialize with NA for those rows that have no gene overlap ###
exon_subset_gene_id_foot <- rep(NA, length(exon_subset_foot))
# Assign the gene_ids where there is a match
exon_subset_gene_id_foot[query_hits_foot] <- gene_ids_foot

### Add gene_id as a metadata column to exon_subset df ###
mcols(exon_subset_foot)$gene_id <- exon_subset_gene_id_foot
exon_subset_gene_id_df_foot<-as.data.frame(exon_subset_foot)
exon_subset_gene_id_df_foot$combined <- paste(exon_subset_gene_id_df_foot$seqnames, exon_subset_gene_id_df_foot$start, sep = "-") #reformat

### Intron subset ###
intron_subset_foot<-subsetByOverlaps(CpGs_foot, intron_GRange)
#intron_subset_foot<-as.data.frame(intron_subset_foot) 

### Adding gene id ###
intron_gff <- subset(all_GRange, type == "intron") #need to be GRange object to subset
mcols(intron_gff)
intron_id_overlaps_foot <- findOverlaps(intron_subset_foot, intron_gff)

### Extract gene_id for matching exons from the GFF file ###
gene_ids_foot<- mcols(intron_gff)$gene[subjectHits(intron_id_overlaps_foot)]
query_hits <- queryHits(intron_id_overlaps_foot)

### Add gene_id only to those rows in exon_subset that have corresponding overlaps ###
### Initialize with NA for those rows that have no overlap ###
intron_subset_gene_id_foot <- rep(NA, length(intron_subset_foot))

### Assign the gene_ids where there is a match ###
intron_subset_gene_id_foot[query_hits] <- gene_ids_foot

### Add gene_id as a metadata column to intron_subset df ###
mcols(intron_subset_foot)$gene_id <- intron_subset_gene_id_foot
intron_subset_gene_id_df_foot<-as.data.frame(intron_subset_foot)
intron_subset_gene_id_df_foot$combined <- paste(intron_subset_gene_id_df_foot$seqnames, intron_subset_gene_id_df_foot$start, sep = "-") #reformat

### Find promoter overlaps with CpG using the adjusted ranges ###
promoter_GRange_foot <- findOverlaps(CpGs_foot, promoters_Grange)

promoter_subset_foot<-subsetByOverlaps(CpGs_foot, promoters_Grange)
promoter_subset_foot_df<-as.data.frame(promoter_subset_foot) 


### Adding gene id ###
promoter_gff <- promoters_Grange # need to be GRange object to subset
promoter_id_overlaps_foot <- findOverlaps(promoter_subset_foot, promoter_gff)

### Extract gene_id for matching promoter region (genes associated with first exons, see above, from the GFF file ###
gene_ids_foot<- mcols(promoter_gff)$gene[subjectHits(promoter_id_overlaps_foot)]
query_hits <- queryHits(promoter_id_overlaps_foot)

### Add gene_id only to those rows in promoter_subset_foot that have corresponding overlaps ###
# Initialize with NA for those rows that have no overlap
promoter_subset_gene_id_foot <- rep(NA, length(promoter_subset_foot))

### Assign the gene_ids where there is a match ###
promoter_subset_gene_id_foot[query_hits] <- gene_ids_foot

### Add gene_id as a metadata column to promoter subset df ###
mcols(promoter_subset_foot)$gene_id <- promoter_subset_gene_id_foot
promoter_subset_gene_id_df_foot<-as.data.frame(promoter_subset_foot)
promoter_subset_gene_id_df_foot$combined <- paste(promoter_subset_gene_id_df_foot$seqnames, promoter_subset_gene_id_df_foot$start, sep = "-") #reformat


### FINDING INTERGENIC REGIONS ###
### Filter for all genic regions (exon, intron, UTRs) ###
non_interg_GRange <- subset(all_GRange, type %in% c("exon", "intron", "five_prime_UTR", "three_prime_UTR"))

### Find overlaps between CpG and non-intergenic regions (genic regions) ###
noninterg_overlap <- findOverlaps(CpGs_foot, non_interg_GRange)

### Get the indices of the CpG regions that overlap with genic regions ###
overlapping_indices <- queryHits(noninterg_overlap)

### Remove these overlapping CpG regions to get intergenic regions ###
interg_subset <- CpGs_foot[-overlapping_indices]

### Convert to data frame and make a Grange Object, genes are NAs ###
interg_subset_df_foot <- as.data.frame(interg_subset) %>% mutate(gene_id =NA)
interg_subset_df_foot$combined <- paste(interg_subset_df_foot$seqnames, interg_subset_df_foot$start, sep = "-")
interg_subset_df_foot_Grange<-makeGRangesFromDataFrame(interg_subset_df_foot)
```


```{r}
############ Running GLM to assess the probability of getting more methylated sites in certain genomic features ############


### Add 1 to all CpG loci to convert positions to 0 base to 1 base format ###
CpGsite_foot<-as.data.frame(y_foot) %>% 
  dplyr::select(Chr, Locus) %>% 
  mutate(Meth=0,
         Locus =Locus +1)

### Reformatting columns to include separate columns for chromosome and locus ###
DM_Trans_foot<- separate(DM_Trans_foot, sample, into = c("Chr", "Locus"), sep = "-")

### DM CpG subset associated with transplant site effect derived before, need to convert to 1 base format too ###
transDM_foot<-DM_Trans_foot%>% 
 mutate(Locus = as.numeric(Locus)+1)

#### Creating a dataframe with all the CpG sites, the ones that are differentially methylated with transplant site were "1" and nonmeth are "0" ###
df_trans <- CpGsite_foot %>%
  left_join(transDM_foot, by = c("Chr", "Locus")) %>%
  mutate(Meth = if_else(!is.na(logFC), 1, Meth)) %>%
  dplyr::select(Chr,Locus, Meth)
  
### Adding genic features to the dataframe for each site ###

intron_subset_foot<-intron_subset_gene_id_df_foot %>% 
  mutate(feature = "intron") #28509 obs

exon_subset_foot<-exon_subset_gene_id_df_foot %>% 
  mutate(feature = "exon") #5694 obs

intergenic_subset_foot<-interg_subset_df_foot %>% 
  mutate(feature = "interg") #44008 obs (this number includes overlap with promoter regions)

promoter_subset_foot<-promoter_subset_gene_id_df_foot %>% 
  mutate(feature = "promoter") #1976 obs

### Finding CpG subsets overlapped with exons ###  
exon<- CpGsite_foot %>%
  full_join(exon_subset_foot, by = c("Chr" = "seqnames", "Locus" = "start")) %>%
na.omit()

### Finding CpG subsets overlapped with introns ###
intron <- CpGsite_foot %>%
  full_join(intron_subset_foot, by = c("Chr" = "seqnames", "Locus" = "start")) %>%
na.omit()

### Finding CpG subsets overlapped with intergenic regions ###
intergenic <- CpGsite_foot %>%
  full_join(intergenic_subset_foot, by = c("Chr" = "seqnames", "Locus" = "start")) %>% 
    filter(!is.na(feature))

### Finding CpG subsets overlapped with promoter regions ###
promoter<-CpGsite_foot %>%
  full_join(promoter_subset_foot, by = c("Chr" = "seqnames", "Locus" = "start")) %>%
na.omit()

### Combining all the CpG subsets ###
df_CpGsite_foot_all <- bind_rows(exon, intron, intergenic, promoter)

### Adding methylation information for foot transplant to the dataframe ###
df_combined_foot_trans <- df_trans%>% 
  full_join(df_CpGsite_foot_all , by =c("Chr","Locus")) %>% 
  dplyr::select(-Meth.y) 

### Remove intergenic regions that overlap with promoter regions ###
  df_combined_foot_trans<-df_combined_foot_trans%>% 
    group_by(Chr, Locus) %>%
 filter(!(any(feature == "promoter") & feature == "interg")) %>%
  ungroup() #intergenic count: 42470 

### Running GLMM to test the effect of genomic feature on DM pattern (0 or 1) ###
### If a CpG falls into intergenic region, it does not have a gene ID and would be "intergenic" 
   df_combined_foot_trans_df <- df_combined_foot_trans%>%
  mutate(gene_id = ifelse(is.na(gene_id), "intergenic", gene_id))

   df_combined_foot_trans_df$feature <-
  as.factor(df_combined_foot_trans_df$feature)
   
### Filter for DM CpGs
   df_combined_foot_trans_df_DM <-df_combined_foot_trans_df %>% 
     filter(Meth.x==1)

glmer <- glmer(Meth.x ~ feature + (1 | gene_id), data =  df_combined_foot_trans_df, family = binomial(link = "logit"))
summary(glmer)
car::Anova(glmer, type ="III",test.statistic="Chisq")

#pairwise comparison
emmeans(glmer, pairwise ~ feature, type="response")

### DM CpGs associated with origin site effect, separating columns into chromosome and locus ###
DM_Origin_foot <- separate(DM_Origin_foot, sample, into = c("Chr", "Locus"), sep = "-")

### Change loci to 0 base to 1 base ###
originDM_foot<-DM_Origin_foot%>% 
 mutate(Locus =as.numeric(Locus)+1)

### Adding methylation information for the CpGs, methylated as 1 and unmethylated as 0 ###
df_origin <- CpGsite_foot %>%
  left_join(originDM_foot, by = c("Chr", "Locus")) %>%
  mutate(Meth = if_else(!is.na(logFC), 1, Meth)) %>%
  dplyr::select(Chr,Locus, Meth)

df_combined_foot_origin <- df_origin %>% 
  full_join(df_CpGsite_foot_all , by =c("Chr","Locus")) %>% 
  dplyr::select(-Meth.y) 

### Remove intergenic regions that overlap with promoter regions ###
df_combined_foot_origin <- df_combined_foot_origin%>% 
    group_by(Chr, Locus) %>%
 filter(!(any(feature == "promoter") & feature == "interg")) %>%
  ungroup()

### Running GLMM to test the effect of feature on DM pattern (0 or 1) ###

### If a CpG falls into intergenic region, it does not have a gene ID and would be "intergenic"
df_combined_foot_origin_df <- df_combined_foot_origin %>%
  mutate(gene_id = ifelse(is.na(gene_id), "intergenic", gene_id))

df_combined_foot_origin_df$feature <-
  as.factor(df_combined_foot_origin_df$feature)

### Filter for DM CpGs
df_combined_foot_origin_df_DM <-df_combined_foot_origin_df %>% 
     filter(Meth.x==1) # no promoters contain DM CpG

### no DM on promoter regions, thus excluding all the promoter entries for the GLM tests
df_combined_foot_origin_df <- df_combined_foot_origin_df %>% filter(feature != "promoter")

glmer <-glmer(Meth.x ~ feature + (1 | gene_id),
              data =  df_combined_foot_origin_df, 
              family = binomial(link = "logit"))
summary(glmer)
car::Anova(glmer, type = "III", test.statistic = "Chisq")

emmeans(glmer, pairwise ~ feature, type="response")

## getting length associated DM CpG information

### DM CpGs associated with length effect, separating columns into chromosome and locus ###
DM_length_foot <- separate(DM_length_foot, sample, into = c("Chr", "Locus"), sep = "-")

### Change loci to 0 base to 1 base ###
lengthDM_foot<-DM_length_foot %>% 
mutate(Locus =as.numeric(Locus)+1)

### Adding methylation information for the CpGs, methylated as 1 and unmethylated as 0 ###
df_length <- CpGsite_foot %>%
  left_join(lengthDM_foot, by = c("Chr", "Locus")) %>%
  mutate(Meth = if_else(!is.na(logFC), 1, Meth)) %>%
  dplyr::select(Chr,Locus, Meth)

df_combined_foot_length <- df_length %>% 
  full_join(df_CpGsite_foot_all , by =c("Chr","Locus")) %>% 
  dplyr::select(-Meth.y) 

### Remove intergenic regions that overlap with promoter regions ###
  df_combined_foot_length <- df_combined_foot_length%>% 
    group_by(Chr, Locus) %>%
 filter(!(any(feature == "promoter") & feature == "interg")) %>%
  ungroup() 

### If a CpG falls into intergenic region, it does not have a gene ID and would be "intergenic"
df_combined_foot_length_df <- df_combined_foot_length %>%
  mutate(gene_id = ifelse(is.na(gene_id), "intergenic", gene_id))  

### Filter for DM CpGs
df_combined_foot_length_DM <- df_combined_foot_length_df %>% 
     filter(Meth.x==1) 


```


```{r}

### Generating table S4 results, DMs associated with features, lfc and GO term information ###

origin_gene_unique <- df_combined_foot_origin %>% filter(Meth.x==1)
transplant_gene_unique <- df_combined_foot_trans %>% filter(Meth.x==1)
length_gene_unique <- df_combined_foot_length %>% filter(Meth.x==1)

#add logfold change info to the DM sites
updated_transplant_gene_unique <- transplant_gene_unique %>% 
  left_join(transDM_foot, by = c("Chr","Locus"))
colnames(updated_transplant_gene_unique)[colnames(updated_transplant_gene_unique) == "logFC"] <- "transplant_logFC"

updated_origin_gene_unique <- origin_gene_unique %>% 
  left_join(originDM_foot, by = c("Chr","Locus")) 
colnames(updated_origin_gene_unique)[colnames(updated_origin_gene_unique) == "logFC"] <- "origin_logFC"

updated_length_gene_unique <- length_gene_unique %>% 
  left_join(lengthDM_foot, by = c("Chr","Locus")) 
colnames(updated_length_gene_unique)[colnames(updated_length_gene_unique) == "logFC"] <- "length_logFC"

gene<-gff_intron %>% 
  filter(type == "gene")

gff_gene_clean <- gene %>%
  separate(attributes, into = c("ID", "Dbxref", "Name", "Description", "Other"), sep = ";", fill = "right") %>%
  mutate(across(everything(), ~ gsub(".*=", "", .))) %>% 
  dplyr::select(Name, Description, start, end)

#want to have gene names, then add the gff_gene_clean data
updated_origin_gene_unique <-left_join(updated_origin_gene_unique ,
                                         gff_gene_clean,
                                         by=c("gene_id"="Name"))
updated_transplant_gene_unique  <-left_join(updated_transplant_gene_unique ,
                                         gff_gene_clean,
                                         by=c("gene_id"="Name"))
updated_length_gene_unique  <-left_join(updated_length_gene_unique ,
                                         gff_gene_clean,
                                         by=c("gene_id"="Name"))

####updated_transplant_gene_unique & updated_origin_gene_unique have all the information of the DM sites, want to see what GO terms those genes were related to.

gene_transplant<-as.data.frame(unique(updated_transplant_gene_unique$gene_id) %>% 
                             na.omit())
gene_transplant$gene_id<-gene_transplant$`unique(updated_transplant_gene_unique$gene_id) %>% na.omit()`


# # load go terms data
go_terms<-
  read.delim(
    "/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/gene ontology/GCF_021869535.1_xbMytCali1.0.p_gene_ontology.gaf",
    header = FALSE,
    sep = "\t",
    # Specify tab delimiter
    comment.char = "!",
    # GAF files have comment lines starting with "!"
    stringsAsFactors = FALSE
  )  %>% # Avoid converting strings to factors)
  dplyr::select(-V11, -V16, -V17)

# Rename columns
colnames(go_terms) <- c("DB", "DB_Object_ID", "gene_id", "Qualifier",
                         "GO_ID", "GO_Term", "Aspect", "DB_Object_Name",
                         "Type", "DB_Object_Type", "Category","Taxon",
                         "V13","V14")

#transplant effect

DMgene_goterm <- gene_transplant %>%
  left_join(go_terms, by = "gene_id") %>%
  group_by(gene_id) %>%
  summarize(GO_terms = list(unique(GO_ID)), .groups = "drop")

DMgene_goterm_flat <- DMgene_goterm %>%
  mutate(across(where(is.list), ~ sapply(., paste, collapse = ";")))


#left joint updated_transplant_gene_unique and the DMgene_goterm_flat to add go term info
updated_transplant_gene_unique<- updated_transplant_gene_unique %>% 
  left_join(DMgene_goterm_flat,
            by = "gene_id")

#write.csv(updated_transplant_gene_unique_foot, "DMgene_goterm_transplant_foot.csv", row.names = FALSE)
######


## Origin site effect
gene_origin<-as.data.frame(unique(updated_origin_gene_unique$gene_id) %>% 
                             na.omit())
gene_origin$gene_id<-gene_origin$`unique(updated_origin_gene_unique$gene_id) %>% na.omit()`
DMgene_goterm <- gene_origin %>%
  left_join(go_terms, by = "gene_id") %>%
  group_by(gene_id) %>%
  summarize(GO_terms = list(unique(GO_ID)), .groups = "drop")

DMgene_goterm_flat <- DMgene_goterm %>%
  mutate(across(where(is.list), ~ sapply(., paste, collapse = ";")))

updated_origin_gene_unique <- updated_origin_gene_unique %>% 
  left_join(DMgene_goterm_flat,
            by = "gene_id")

#write.csv(updated_origin_gene_unique, "DMgene_goterm_origin_foot.csv", row.names = FALSE)

###length effect

gene_length<-as.data.frame(unique(updated_length_gene_unique$gene_id) %>% 
                             na.omit())
gene_length$gene_id<-gene_length$`unique(updated_length_gene_unique$gene_id) %>% na.omit()`

DMgene_goterm <- gene_length %>%
  left_join(go_terms, by = "gene_id") %>%
  group_by(gene_id) %>%
  summarize(GO_terms = list(unique(GO_ID)), .groups = "drop")

DMgene_goterm_flat <- DMgene_goterm %>%
  mutate(across(where(is.list), ~ sapply(., paste, collapse = ";")))

updated_length_gene_unique <- updated_length_gene_unique %>% 
  left_join(DMgene_goterm_flat,
            by = "gene_id")

#write.csv(updated_length_gene_unique, "DMgene_goterm_length_foot.csv", row.names = FALSE)

#look at how many DM CpGs were shared between transplant and origin site effect
origin_combined <- unique(updated_origin_gene_unique$combined)
transplant_combined <- unique(updated_transplant_gene_unique$combined)
shared_combined <- intersect(origin_combined, transplant_combined)
# Count how many are shared
length(shared_combined)
#24


#### Making input file for t-test ###
# Ensure all three data frames have the same column names
colnames(updated_origin_gene_unique) <- colnames(updated_transplant_gene_unique)
#colnames(updated_transplant_gene_unique) <- colnames(updated_length_gene_unique)
merged_all_DM_foot <- rbind(
                    updated_origin_gene_unique,
                    updated_transplant_gene_unique)

#look at whether strengh of DM differ between the two effects in foot
trans_origin_compar_foot<-lmer(abs(transplant_logFC)~Treat+(1|gene_id),merged_all_DM_foot)
summary(trans_origin_compar_foot)

#write_csv(merged_all_DM_foot,"/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/manuscript/merged_all_DM_foot.csv")
# 
# average_lfc_DM_trans_foot<-updated_transplant_gene_unique %>%
#   group_by(gene_id) %>%
#   summarize(mean = mean(transplant_logFC))%>%
#   na.omit()
# 
# average_lfc_DM_origin_foot<-updated_origin_gene_unique %>%
#   group_by(gene_id) %>%
#   summarize(mean = mean(transplant_logFC))%>%
#   na.omit()
# 
#  write_csv(average_lfc_DM_trans_foot,"/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/MethylMussel_GO/GO_MWU/average_lfc_DM_foot_trans.csv")
# 
#   write_csv(average_lfc_DM_origin_foot,"/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/MethylMussel_GO/GO_MWU/average_lfc_DM_foot_origin.csv")
# 

```




```{r}

############ GO Enrichment analysis ############
library(goseq)
### Load in GO terms, downloaded from NCBI ###
go_terms<-
  read.delim(
    "/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/gene ontology/GCF_021869535.1_xbMytCali1.0.p_gene_ontology.gaf",
    header = FALSE,
    sep = "\t",
    # Specify tab delimiter
    comment.char = "!",
    # GAF files have comment lines starting with "!"
    stringsAsFactors = FALSE
  )  %>% # Avoid converting strings to factors) 
  dplyr::select(-V11, -V16, -V17)

### Changing column names ###
  colnames(go_terms) <- c("DB", "DB_Object_ID", "gene_id", "Qualifier", 
                         "GO_ID", "GO_Term", "Aspect", "DB_Object_Name", 
                         "Type", "DB_Object_Type", "Category","Taxon", 
                         "V13","V14") 

### Changing DM CpG to 1 base format ###
transDM_foot<-DM_Trans_foot %>% 
  mutate(Locus=as.numeric(Locus)+1)

originDM_foot<- DM_Origin_foot %>% 
  mutate(Locus=as.numeric(Locus)+1)

### Creating a dataframe with genomic feature & DM methylation classification, keep only entres with genes ###
df_trans_all_fix <- df_CpGsite_foot_all %>%
  left_join(originDM_foot, by = c("Chr", "Locus")) %>%
  mutate(Meth = if_else(!is.na(logFC), 1, Meth)) %>%
  dplyr::select(Chr,Locus, Meth,gene_id,feature) %>%  # Removing intergenic features that do not correspond with a gene
  na.omit()

### code for origin site effect ###
# df_trans_all_fix <- df_CpGsite_foot_all %>%
#   left_join(originDM_foot, by = c("Chr", "Locus")) %>%
#   mutate(Meth = if_else(!is.na(logFC), 1, Meth)) %>%
#   dplyr::select(Chr,Locus, Meth,gene_id,feature) %>%  
#   na.omit()

### Extracting all DM CpGs ###
  DM_direc <- df_trans_all_fix %>%
  filter(Meth == 1) %>%
  na.omit()

### Extracting DM CpGs associated with both transplant and origin site effect ###
transplant_feature<-df_combined_foot_trans %>% filter(Meth.x ==1)%>%
  na.omit()

origin_feature<-df_combined_foot_origin %>% filter(Meth.x ==1)%>%
  na.omit()


DM_direc_origin<-left_join(DM_direc,
                           origin_feature,
                           by=c("gene_id","Locus","Chr"))

DM_direc_transplant<-left_join(DM_direc,
                            transplant_feature,
                            by=c("gene_id","Locus","Chr"))


### Matching genes with go terms ###
go_foot_trans_dm <- df_trans_all_fix %>% 
  left_join(go_terms, by = c("gene_id")) %>%  
  na.omit() 

### Count unique CpG sites per gene ###
cpg_per_gene<-go_foot_trans_dm  %>% 
  group_by(gene_id) %>% 
 summarize(unique_locus_count = n_distinct(paste(Locus, Chr))) %>% 
  na.omit()

### Extract unique gene IDs from GO terms ###
gene_id<-go_foot_trans_dm %>% dplyr::select(gene_id) %>% 
  distinct()


### Filter out genes that have CpG coverage less than 3, save as cpg_sites_bias ###
cpg_sites_bias<- gene_id %>%
  left_join(cpg_per_gene, by = c("gene_id")) %>% 
  filter(unique_locus_count>=3) 

### GO Matching category file, which contains gene IDs and their corresponding GO IDs ###
category_file<-go_foot_trans_dm %>% 
  dplyr::select(gene_id, GO_ID) %>% 
  distinct()%>%
  filter(gene_id %in% cpg_sites_bias$gene_id)%>% 
   distinct(gene_id, GO_ID)


############ For all GO TERMS extract the Parental level ############

### Loading libraries ###
library(GO.db)
library(AnnotationDbi) 

### Extract all the unique GO terms from category list ###
go_terms_cat <- unique(category_file$GO_ID)

### Extract ontology for each GO terms ###
go_info <- AnnotationDbi::select(GO.db, keys = go_terms_cat, columns = c("ONTOLOGY", "TERM"))

### Creating separate dataframe to store BP, MF, CC GO terms ###
bp<- go_info %>% 
  filter(ONTOLOGY == "BP") 
mf<-go_info %>% 
  filter(ONTOLOGY == "MF")
cc<-go_info %>% 
  filter(ONTOLOGY == "CC")

### BIOLOGICAL PROCESS ###
# Initialize a list to store parent terms
parent_terms <- list()

# Loop through each GO term and get the parent
for (go in bp$GOID) {
  # Attempt to retrieve the parent term (one level up)
  parent <- as.character(GO.db::GOBPPARENTS[[go]])
  
  # Debugging output
  if (is.null(parent) || length(parent) == 0) {
    cat("No parent term found for:", go, "\n")
    parent_terms[[go]] <- go  # Keep the original GO term if no parent is found
  } else {
    parent_terms[[go]] <- parent
  }
}

BP_parent_terms_df <- data.frame(
  GO_Term = rep(bp$GOID, sapply(parent_terms, length)),
  Parent_Term = unlist(parent_terms, use.names = FALSE),
  stringsAsFactors = FALSE) %>% 
  mutate(Ontology ="BP")


### CELLULAR COMPONENT ###
# Initialize a list to store parent terms
parent_terms <- list()

# Loop through each GO term and get the parent
for (go in cc$GOID) {
  # Attempt to retrieve the parent term
  parent <- as.character(GO.db::GOCCPARENTS[[go]])
  
  # Debugging output
  if (is.null(parent) || length(parent) == 0) {
    cat("No parent term found for:", go, "\n")
    parent_terms[[go]] <- go  # Keep the original GO term if no parent is found
  } else {
    parent_terms[[go]] <- parent
  }
}

CC_parent_terms_df <- data.frame(
  GO_Term = rep(cc$GOID, sapply(parent_terms, length)),
  Parent_Term = unlist(parent_terms, use.names = FALSE),
  stringsAsFactors = FALSE)%>% 
  mutate(Ontology ="CC")


  
### MOLECULAR FUNCTION ###
# Initialize a list to store parent terms
parent_terms <- list()

# Loop through each GO term and get the parent
for (go in mf$GOID) {
  # Attempt to retrieve the parent term
  parent <- as.character(GO.db::GOMFPARENTS[[go]])
  
  # Debugging output
  if (is.null(parent) || length(parent) == 0) {
    cat("No parent term found for:", go, "\n")
    parent_terms[[go]] <- go  # Keep the original GO term if no parent is found
  } else {
    parent_terms[[go]] <- parent
  }
}

MF_parent_terms_df <- data.frame(
  GO_Term = rep(mf$GOID, sapply(parent_terms, length)),
  Parent_Term = unlist(parent_terms, use.names = FALSE),
  stringsAsFactors = FALSE) %>% mutate(Ontology ="MF")


### Combine the BP, MF, CC data frames into one ###
all_parent_terms_df <- bind_rows(MF_parent_terms_df, CC_parent_terms_df, BP_parent_terms_df)

#### Rejoin with category file to get all the gene ID that matches those parental terms ###

category_file_parent<-category_file %>% 
  left_join(all_parent_terms_df, by=c("GO_ID" ="GO_Term"))

############################################################

### Only keep GO terms with at least 3 genes ###
gene_go_parent<-category_file_parent %>%
  group_by(Parent_Term) %>%
summarise(gene_count = n_distinct(gene_id)) %>% ###IMPORTANT CHANGE: change to n_distinct
  filter(gene_count >= 3)

### Removing old GO terms, only including the parental GO terms ###
category_list_prep_parent <- gene_go_parent %>%
  left_join(category_file_parent, by = "Parent_Term") %>% 
  dplyr::select(-GO_ID)%>% 
   distinct(Parent_Term, gene_id, .keep_all = TRUE)

### Filtering out CpsG, include only those included in cpg_sites_bias (See above, only include genes with with at least 3 CpG coverage) ###
category_list <- cpg_sites_bias %>% 
  left_join(category_list_prep_parent, by = "gene_id") %>%  # Keep only genes in cpg_sites_bias
  na.omit() %>% 
  group_by(gene_id) %>%
  summarise(GO_ID = list(unique(Parent_Term)), .groups = 'drop') %>%  # Drop grouping afterwards
  deframe()


# Build 2-col gene parent mapping
gene2cat_df <- category_file_parent %>%         # (gene_id, GO_ID, Parent_Term) from your code
  filter(!is.na(Parent_Term)) %>%
  count(Parent_Term, name = "gene_count") %>%
  filter(gene_count >= 3) %>%
  dplyr::select(Parent_Term) %>%
  inner_join(category_file_parent, by = "Parent_Term") %>%
  transmute(gene = as.character(gene_id),
                   category = as.character(Parent_Term)) %>%
  distinct()


### The input bias correction file, which contains the gene id and the number of CpGs for that gene ###
cpg_sites_bias_parent<- category_list_prep_parent  %>% 
  left_join(cpg_sites_bias, by ="gene_id") %>% 
   dplyr::select(gene_id, unique_locus_count) %>% 
    distinct(gene_id, unique_locus_count) 

### Reformat bias correction file to vector format ###
cpg_vector <- setNames(cpg_sites_bias_parent$unique_locus_count,
                          cpg_sites_bias_parent$gene_id)

### Create dataframe with gene id, CpG count for each locus, methylation information, and GO annotations ###
go_foot_trans_dm_parent <- cpg_sites_bias_parent %>% 
  left_join(go_foot_trans_dm, by ="gene_id")

  
### Separating unmethylated and methylated CpGs ###
DMG<-go_foot_trans_dm_parent%>% 
  filter(Meth ==1) %>% 
   dplyr::select(gene_id, GO_ID) %>% 
  distinct()

UD<-go_foot_trans_dm_parent %>% 
  filter(Meth ==0) %>% 
   dplyr::select(gene_id, GO_ID) %>% 
  distinct()

### extract DM and Unmeth gene ID ###
dm_genes <- DMG$gene_id
ud_genes <- UD$gene_id

### Combine all genes into a single vector ###
all_genes <- union(dm_genes, ud_genes)

# Start with everything as 0
gene_vector <- setNames(rep(0, length(all_genes)), all_genes)

# Overwrite DM genes as 1
gene_vector[dm_genes] <- 1

table(gene_vector) 

# Enforce identical universe & order
common_ids <- Reduce(intersect, list(names(gene_vector), names(cpg_vector), gene2cat_df$gene))
gene_vector <- gene_vector[common_ids]
cpg_vector  <- cpg_vector[common_ids]
gene2cat_df <- filter(gene2cat_df, gene %in% common_ids)


### Running GO enrichment analysis with bias data ###
pwf <- nullp(gene_vector, bias.data=cpg_vector, plot.fit =FALSE)
head(pwf)

plotPWF(pwf = pwf, binsize = 200) 

GO.wall <- goseq(pwf, gene2cat = gene2cat_df, method = "Wallenius", use_genes_without_cat = FALSE)
# adjust each direction separately
GO.wall$padj_over  <- p.adjust(GO.wall$over_represented_pvalue,  method = "BH")
GO.wall$padj_under <- p.adjust(GO.wall$under_represented_pvalue, method = "BH")

overrep_go<-GO.wall %>% filter(over_represented_pvalue<0.05)
enriched_go_ids <- overrep_go$category

 ### FDR corrected ###
enriched.GO <- GO.wall$category[p.adjust(GO.wall$over_represented_pvalue, method = "BH" ) < 0.05]
 head(enriched.GO)
 #character(0)
 
 write_csv(GO.wall, "goseq_origin.csv")

```


```{r}

#prep for GOMU test

all_CpG_dm_foot$Chr <- sapply(strsplit(as.character(all_CpG_dm_foot$sample), "-"), `[`, 1)  # Part before '-'
all_CpG_dm_foot$Locus <- sapply(strsplit(as.character(all_CpG_dm_foot$sample), "-"), `[`, 2)
all_CpG_dm_foot$Locus<-as.numeric(all_CpG_dm_foot$Locus)+1


origin_gene_unique_all <- all_CpG_dm_foot %>% 
  filter(Treat=="Origin")

transplant_gene_unique_all <- all_CpG_dm_foot %>% 
  filter(Treat=="Transplant")


#add logfold change info to the DM sites
updated_transplant_gene_unique_all <- df_combined_foot_trans  %>% 
  left_join(transplant_gene_unique_all, by = c("Chr","Locus"))
colnames(updated_transplant_gene_unique_all)[colnames(updated_transplant_gene_unique_all) == "logFC"] <- "transplant_logFC"

updated_origin_gene_unique_all <- df_combined_foot_origin %>% 
  left_join(origin_gene_unique_all, by = c("Chr","Locus")) 
colnames(updated_origin_gene_unique_all)[colnames(updated_origin_gene_unique_all) == "logFC"] <- "origin_logFC"


#want to have gene names, then add the gff_gene_clean data
updated_origin_gene_unique_all <-left_join(updated_origin_gene_unique_all ,
                                         gff_gene_clean,
                                         by=c("gene_id"="Name"))
updated_transplant_gene_unique_all  <-left_join(updated_transplant_gene_unique_all ,
                                         gff_gene_clean,
                                         by=c("gene_id"="Name"))

####updated_transplant_gene_unique & updated_origin_gene_unique have all the information of the DM sites, want to see what GO terms those genes were related to.

gene_transplant_all<-as.data.frame(unique(updated_transplant_gene_unique_all$gene_id) %>% 
                             na.omit())
gene_transplant_all$gene_id<-gene_transplant_all$`unique(updated_transplant_gene_unique_all$gene_id) %>% na.omit()`



#transplant effect

DMgene_goterm_all <- gene_transplant_all %>%
  left_join(go_terms, by = "gene_id") %>%
  group_by(gene_id) %>%
  summarize(GO_terms = list(unique(GO_ID)), .groups = "drop")

DMgene_goterm_flat_all <- DMgene_goterm_all %>%
  mutate(across(where(is.list), ~ sapply(., paste, collapse = ";")))


#left joint updated_transplant_gene_unique and the DMgene_goterm_flat to add go term info
updated_transplant_gene_unique_all<- updated_transplant_gene_unique_all %>% 
  left_join(DMgene_goterm_flat_all,
            by = "gene_id")

#write.csv(updated_transplant_gene_unique_foot, "DMgene_goterm_transplant_foot.csv", row.names = FALSE)
######


## Origin site effect
gene_origin_all<-as.data.frame(unique(updated_origin_gene_unique_all$gene_id) %>% 
                             na.omit())
gene_origin_all$gene_id<-gene_origin_all$`unique(updated_origin_gene_unique_all$gene_id) %>% na.omit()`
DMgene_goterm_all <- gene_origin_all %>%
  left_join(go_terms, by = "gene_id") %>%
  group_by(gene_id) %>%
  summarize(GO_terms = list(unique(GO_ID)), .groups = "drop")

DMgene_goterm_flat_all <- DMgene_goterm_all %>%
  mutate(across(where(is.list), ~ sapply(., paste, collapse = ";")))

updated_origin_gene_unique_all <- updated_origin_gene_unique_all %>% 
  left_join(DMgene_goterm_flat_all,
            by = "gene_id")

#write.csv(updated_origin_gene_unique, "DMgene_goterm_origin_foot.csv", row.names = FALSE)

## Getting average LFC  ###
average_lfg_foot_origin<-updated_origin_gene_unique_all %>% 
  group_by(gene_id) %>% 
  summarize(mean=mean(origin_logFC)) %>% 
  na.omit()

average_lfg_foot_transplant<-updated_transplant_gene_unique_all %>% 
  group_by(gene_id) %>% 
  summarize(mean=mean(transplant_logFC)) %>% 
  na.omit()


go_foot_df <- category_list_prep_parent %>% group_by(gene_id) %>%
  summarise(GO = paste(unique(Parent_Term), collapse = ";"), .groups = "drop")


### Average LFC after applying filtering threshold ###

lfg_gene_foot_origin_filtered<-updated_origin_gene_unique_all %>% 
  filter(gene_id %in% go_foot_df$gene_id)

average_lfg_foot_origin_filtered<-lfg_gene_foot_origin_filtered%>% 
  group_by(gene_id) %>% 
  summarize(mean=mean(origin_logFC)) %>% 
  na.omit()

lfg_gene_foot_transplant_filtered<-updated_transplant_gene_unique_all %>% 
  filter(gene_id %in% go_foot_df$gene_id)

average_lfg_foot_transplant_filtered<-lfg_gene_foot_transplant_filtered %>% 
  group_by(gene_id) %>% 
  summarize(mean=mean(transplant_logFC)) %>% 
  na.omit()


#save tye average log fold change values per gene

# write_csv(average_lfg_foot_transplant_filtered,"/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/MethylMussel_GO/GO_MWU/average_lfg_foot_transplant_filtered_no_interaction.csv")
# 
# write_csv(average_lfg_foot_origin_filtered,"/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/MethylMussel_GO/GO_MWU/average_lfg_foot_origin_filtered_no_interaction.csv")


# ## saving the matching GO term and gene file ###
# write_csv(category_list_prep_parent,"/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/MethylMussel_GO/GO_MWU/go_gene_matching.csv")

go_foot_df <-category_list_prep_parent %>%
  group_by(gene_id) %>%
  summarise(GO = paste(unique(Parent_Term), collapse = ";"), .groups = "drop")

# write.table(
#   go_foot_df,
#   "/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/MethylMussel_GO/GO_MWU/GO_foot_df.tab",
#   row.names = FALSE,
#   sep = "\t",
#   quote = FALSE,
#   col.names = FALSE
# )
# 


```