```{r}
### Set working directory, include all (top5 for each treatments) gill coverage files ###
setwd("/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/cg_coverage_files/tissue_specific/gill")

### Load metadata file that include sample names, load all coverage files into R ###
meta_data_gill<-read.delim("gill_metadata.txt", row.names = "sample", stringsAsFactors = FALSE)
Sample_gill <- row.names(meta_data_gill)
files_gill <- paste0(Sample_gill,".CpG_report.merged_CpG_evidence.cov.CpG_report.merged_CpG_evidence.cov")

### EdgeR function readBismark2DGE reads all the files and collates the counts for all the sample into one data object ###
yall <- readBismark2DGE(files_gill, sample.names=Sample_gill)

### Check dimension and the count matrix ###
dim(yall)
head(yall$counts)

### Save a data frame for downstream analyses ###
yall_df<-as.data.frame(yall)

### Sum up the counts of methylated and unmethylated reads to get the total read coverage at each CpG site for each sample ###
Methylation <- gl(2, 1, ncol(yall), labels = c("Me", "Un"))
Me <- yall$counts[ , Methylation == "Me" ]
Un <- yall$counts[ , Methylation == "Un" ]
Coverage <- Me + Un
# Prefiltered total # of CpGs sequenced: colSums(Coverage > 0, na.rm = TRUE) 
head(Coverage)

### Calculating prefiltered genome-wide coverage - Replace zeros with NA so they're excluded in mean
# Coverage[Coverage == 0] <- NA
# # Compute mean per column, excluding NA (formerly zero)
# mean_coverage <- colMeans(Coverage, na.rm = TRUE)
# mean_coverage

### Calculating average genome wide methylation percentage prefiltering
# prop_meth_matrix_gill_prefilter <- Me /(Un+Me)
# colMeans(prop_meth_matrix_gill_prefilter,na.rm=TRUE)*100

## Calculating # of unique CpGs 
# covered_per_cpg <- rowSums(Coverage > 0, na.rm = TRUE)
# sum(covered_per_cpg > 0) # 1946085

### Filtering to only include samples with a coverage of at least 3, for 66% of samples, total of 14 samples ###
n=3
keep_gill <- rowSums(Coverage >= n) >= 14
table(keep_gill)

### DGEList object is subsetted to retain only the filtered loci ###
y_gill <- yall[keep_gill,, keep.lib.sizes=FALSE]

### Normalization - set the library sizes for each sample to be the average of the total read counts for the methylated and unmethylated libraries ###
TotalLibSize <- y_gill$samples$lib.size[Methylation=="Me"] +
  +                 y_gill$samples$lib.size[Methylation=="Un"]
y_gill$samples$lib.size <- rep(TotalLibSize, each=2)
y_gill$samples

### Compute the corresponding methylation summary from the methylated and unmethylated counts ###
Me_gill <- y_gill$counts[, Methylation=="Me"]
Un_gill <- y_gill$counts[, Methylation=="Un"]

### Calculating prefiltered genome-wide coverage 
# Coverage_postfilter<-Un_gill + Me_gill
# Coverage_postfilter[Coverage_postfilter == 0] <- NA
# mean_coverage_postfilter <- colMeans(Coverage_postfilter, na.rm = TRUE)
# mean_coverage_postfilter 

### Calculating a methylation proportion matrix
prop_meth_matrix_gill <- Me_gill/(Me_gill+Un_gill)
#Postfiltered total # of CpGs: colSums(!is.na(prop_meth_matrix_gill))

# ### Calculating average genome wide methylation percentage postfiltering
# colMeans(prop_meth_matrix_gill,na.rm=TRUE)*100

```


```{r}

### We want to use the exposed site as reference level for both transplant and origin site effect analyses, need to relevel for transplant site effect ###
meta_data_gill$transplant_site <- relevel(factor(meta_data_gill$transplant_site), ref = "exposed")
meta_data_gill$origin_site <- relevel(factor(meta_data_gill$origin_site), ref = "exposed")

meta_data_gill$scale_len <- scale(meta_data_gill$LENGTH_FINAL..mm.)

### Create a design matrix, using origin site, transplant site, and final shell length as fixed effects ###
designSL_gill <-
  model.matrix( ~ 1 + origin_site * transplant_site + scale_len,
                data = meta_data_gill)

### Expand to the full design matrix modeling the sample and methylation effects ###
design_gill <- modelMatrixMeth(designSL_gill)

### Dispersion estimation ###
y_gill <- estimateDisp(y_gill, design = design_gill, robust = TRUE)

### Create the BCV plot for visualization ### 
plotBCV(y_gill)

### Testing for differentially methylated CpG loci ###

### fit NB GLMs for all the CpG loci using the glmFit function in edgeR ###
fit_gill <- glmFit(y_gill, design_gill)
```


```{r}

###getting origin: transplant intearction
interaction_terms <- grep("origin_site.*:transplant_site.*", colnames(fit_gill$design))

lrt_gill_interaction <- glmLRT(fit_gill, coef = interaction_terms)
summary( decideTests(lrt_gill_interaction) )
plotMD(lrt_gill_interaction)

#remove colons to reformat the names
colnames(design_gill) <- make.names(colnames(design_gill))

###Origin site effect####
contr_origin_gill <- makeContrasts(Origin = origin_siteprotected, levels = design_gill)
lrt_origin_gill <- glmLRT(fit_gill, contrast=contr_origin_gill)
summary( decideTests(lrt_origin_gill) )
plotMD(lrt_origin_gill)

###Transplant effect ####
contr_trans_gill <- makeContrasts(Transplant = transplant_siteprotected, levels = design_gill)
lrt_trans_gill <- glmLRT(fit_gill, contrast=contr_trans_gill)
summary( decideTests(lrt_trans_gill) )
plotMD(lrt_trans_gill)


## Wrangle data for volano plot of CpG diff meth
# Correct p-values
lrt_origin_gill$table$FDR <- p.adjust( lrt_origin_gill$table$PValue, method = "BH" )
lrt_trans_gill$table$FDR <- p.adjust( lrt_trans_gill$table$PValue, method = "BH" )
lrt_gill_interaction$table$FDR <- p.adjust( lrt_gill_interaction$table$PValue, method = "BH" )
lrt_gill_interaction$table$FDR <- p.adjust(lrt_gill_interaction$table$PValue, method = "BH" )


# Apply logical variable for significant DM
lrt_origin_gill$table$Sig <- ifelse(lrt_origin_gill$table$FDR < 0.05, TRUE, FALSE )
lrt_trans_gill$table$Sig <- ifelse(lrt_trans_gill$table$FDR < 0.05, TRUE, FALSE )
lrt_gill_interaction$table$Sig <- ifelse(lrt_gill_interaction$table$FDR < 0.05, TRUE, FALSE )
lrt_gill_interaction$table$Sig <- ifelse(lrt_gill_interaction$table$FDR < 0.05, TRUE, FALSE )

# Apply binary variable for hyper/hypomethylation or "Up" vs "Down"
lrt_origin_gill$table$Dir <- ifelse(lrt_origin_gill$table$logFC > 0, "Up", "Down" )
lrt_trans_gill$table$Dir <- ifelse( lrt_trans_gill$table$logFC > 0, "Up", "Down" )
lrt_gill_interaction$table$Dir <- ifelse( lrt_gill_interaction$table$logFC > 0, "Up", "Down" )
lrt_gill_interaction$table$Dir <- ifelse(lrt_gill_interaction$table$logFC > 0, "Up", "Down")



# Create combined term for significance and fold-change direction of diff meth
lrt_origin_gill$table$Sig_Dir <- paste( lrt_origin_gill$table$Sig,
                             lrt_origin_gill$table$Dir,
                             sep = "_" )
lrt_trans_gill$table$Sig_Dir <- paste( lrt_trans_gill$table$Sig,
                             lrt_trans_gill$table$Dir,
                             sep = "_" )
lrt_gill_interaction$table$Sig_Dir <- paste( lrt_gill_interaction$table$Sig,
                             lrt_gill_interaction$table$Dir,
                             sep = "_" )
lrt_gill_interaction$table$Sig_Dir <- paste( lrt_gill_interaction$table$Sig,
                             lrt_gill_interaction$table$Dir,
                             sep = "_" )


# Create CpG ID and treatment variables
lrt_origin_gill$table$Treat <- "Origin"
lrt_trans_gill$table$Treat <- "Transplant"
lrt_gill_interaction$table$Treat <- "Interaction"


lrt_origin_gill$table$sample <- rownames(lrt_origin_gill$table)
lrt_trans_gill$table$sample <- rownames(lrt_trans_gill$table)
lrt_gill_interaction$table$sample <- rownames(lrt_gill_interaction$table)
lrt_gill_interaction$table$sample <- rownames(lrt_gill_interaction$table)


# Merge  coefficients
all_CpG_dm <- rbind( lrt_origin_gill$table,
                     lrt_trans_gill$table,
                     lrt_gill_interaction$table)
write_csv(all_CpG_dm, "/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/manuscript/gill_interaction_all_CpG_dm.csv")

gill_all_CpG_dm<-rbind( lrt_origin_gill$table,
                     lrt_trans_gill$table,
                     lrt_gill_interaction$table)


DM_Trans_gill<-all_CpG_dm %>% filter(Sig =="TRUE",Treat =="Transplant")
DM_Origin_gill<-all_CpG_dm %>% filter(Sig =="TRUE", Treat =="Origin")
DM_interaction_gill<-all_CpG_dm %>% filter(Sig =="TRUE", Treat =="Interaction")
  
DM_gill<-all_CpG_dm %>% filter(Sig =="TRUE", Treat =="Length")
DM_interact_gill<-all_CpG_dm %>% filter(Sig =="TRUE", Treat =="Length_interact")

#  volcano plots
gill_dff_meth<-ggplot( data = all_CpG_dm, aes(y = -log(PValue), x = as.numeric( logFC ),
                               color = Sig_Dir ) ) +
    theme_classic(base_size = 40) +
    geom_point(size=4, alpha =0.5, stroke=0) +
    theme( legend.position = "none",
           strip.background = element_blank() ) +
    scale_color_manual( values = c( "black", "black", "blue", "red" ) ) +
    facet_grid( Treat ~ . , scale = "free") +
     labs(x = "log2FC", y = "-log p-value") +
  theme(
  axis.text.x = element_text(size = 12),
  axis.ticks.x = element_line()
)

gill_dff_meth

```

```{r}
library(patchwork)
# Filter datasets
plot_origin <- all_CpG_dm %>% filter(Treat == "Origin")
plot_trans <- all_CpG_dm %>% filter(Treat == "Transplant")
plot_interaction<- all_CpG_dm %>% filter(Treat == "Interaction")
  
  
# Individual plots
p1_gill <- ggplot(plot_trans, aes(x = logFC, y = -log10(PValue), color = Sig_Dir)) +
  geom_point(size = 5, alpha =0.4, stroke=0) +
  scale_color_manual(values = c("#2a2e2e", "#2a2e2e", "blue", "red")) +
  labs(x = "", y = "", title = "Transplant",color="black", size =28) +
  theme_few(base_size = 28) +
  theme(legend.position = "none",
        plot.title = element_blank(),
   axis.text = element_text(color = "black"),
    axis.ticks = element_line(color = "black"),
    axis.title = element_text(color = "black")
   )

p1_gill 

p2_gill <- ggplot(plot_origin, aes(x = logFC, y = -log10(PValue), color = Sig_Dir))+
  geom_point(size = 5, alpha =0.4,stroke=0) +
  scale_color_manual(values = c("#2a2e2e", "#2a2e2e", "blue", "red")) +
  labs(x = "", y = "", title = "Origin", color="black", size =28) +
  theme_few(base_size = 28) +
  theme(legend.position = "none",
        plot.title = element_blank(),
   axis.text = element_text(color = "black"),
    axis.ticks = element_line(color = "black"),
    axis.title = element_text(color = "black"))
p2_gill 

p3_gill <- ggplot(plot_interaction, aes(x = logFC, y = -log10(PValue), color = Sig_Dir))+
  geom_point(size = 5, alpha =0.4,stroke=0) +
  scale_color_manual(values = c("#2a2e2e", "#2a2e2e", "blue", "red")) +
  labs(x = "log2FC", y = "", title = "Interaction", color="black", size =28) +
  theme_few(base_size = 28) +
  theme(legend.position = "none",
        plot.title = element_blank(),
   axis.text = element_text(color = "black"),
    axis.ticks = element_line(color = "black"),
    axis.title = element_text(color = "black"))

p3_gill


common_xlim <- c(-15, 15)
common_ylim <- c(0, 12)
# aspect_ratio_value <- 0.6  # Adjust this value as needed (1 means square)


p1_gill <- p1_gill + coord_cartesian(xlim = common_xlim, ylim = common_ylim)
p2_gill <- p2_gill + coord_cartesian(xlim = common_xlim, ylim = common_ylim)
p3_gill <- p3_gill + coord_cartesian(xlim = common_xlim, ylim = common_ylim)

# Stack vertically with shared layout
gill_volcano<-p2_gill / p1_gill / p3_gill  # This uses patchwork
gill_volcano

#ggsave("gill_volcano_interaction.png", gill_volcano, width = 10, height = 12, dpi = 300)


print(gill_volcano + plot_layout(heights = c(1, 1, 1)))  # ensure equal height per row

```


```{r}

### making combined tissue volcano plots for effects without length

foot_volcano_labeled <- foot_volcano + plot_annotation(title = "A. Foot",
                                                       theme = theme(plot.title = element_text(face= "bold", size = 30)))
gill_volcano_labeled <- gill_volcano + plot_annotation(title = "B. Gill",
                                                       theme = theme(plot.title = element_text(face= "bold", size = 30)))

# 2. Combine them without automatic tags
combined <- wrap_elements(foot_volcano_labeled) + wrap_elements(gill_volcano_labeled) +
  plot_layout(ncol = 2, guides = "collect") +
  plot_annotation(tag_levels = NULL)  # turn off A/B tags

combined<- combined + plot_layout(guides = "collect") & 
  theme(plot.margin = margin(t = 5, r = 40, b = 5, l = 5))  # Add right padding

combined

# png("/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/manuscript/submission/DM_combined.png", width = 23, height = 15, units = "in", res = 300)
png("/Users/qcai/Documents/UCSC/Kelley_Lab/mytilus/manuscript/M_combined.png", width = 28, height = 27, units = "in", res = 300)

print(combined)



# Add "Origin" to top-right
grid.text("Origin", x = unit(0.96, "npc"), y = unit(0.84, "npc"),
          rot = 270, gp = gpar(fontsize = 26, fontface = "bold"))

# Add "Transplant" to middle -right
grid.text("Transplant", x = unit(0.96, "npc"), y = unit(0.51, "npc"),
          rot = 270, gp = gpar(fontsize = 26, fontface = "bold"))

# Add "Interaction" to bottom-right
grid.text("Interaction", x = unit(0.96, "npc"), y = unit(0.19, "npc"),
          rot = 270, gp = gpar(fontsize = 26, fontface = "bold"))


# 4. Close the device to save the file
dev.off()
```